File: ICSharpCode.Decompiler.Tests/TestCases/VBPretty/VBTryCatchFinally.vb
Imports System
Imports System.IO

Public Class VBTryCatchFinally
	Private Shared Function Condition() As Boolean
		Return True
	End Function

	Public Shared Sub TryFinally()
		Try
			Console.WriteLine("Try")
		Finally
			Console.WriteLine("Finally")
		End Try
	End Sub

	Public Shared Sub TryCatchBare()
		Try
			Console.WriteLine("Try")
		Catch
			Console.WriteLine("Catch")
		End Try
	End Sub

	Public Shared Sub TryCatchVariable()
		Try
			Console.WriteLine("Try")
		Catch ex As Exception
			Console.WriteLine(ex.Message)
		End Try
	End Sub

	Public Shared Sub TryCatchSpecificType()
		Try
			Console.WriteLine("Try")
		Catch ex As IOException
			Console.WriteLine(ex.Message)
		End Try
	End Sub

	Public Shared Sub TryCatchUnusedVariable()
		Try
			Console.WriteLine("Try")
		Catch ex As InvalidOperationException
			Console.WriteLine("Catch")
		End Try
	End Sub

	Public Shared Sub TryCatchWhen()
		Try
			Console.WriteLine("Try")
		Catch When Condition()
			Console.WriteLine("Catch")
		End Try
	End Sub

	Public Shared Sub TryCatchVariableWhen()
		Try
			Console.WriteLine("Try")
		Catch ex As Exception When ex.Message IsNot Nothing
			Console.WriteLine(ex.Message)
		End Try
	End Sub

	Public Shared Sub TryCatchSpecificTypeWhen()
		Try
			Console.WriteLine("Try")
		Catch ex As IOException When Condition()
			Console.WriteLine(ex.Message)
		End Try
	End Sub

	Public Shared Sub TryCatchExistingLocal()
		Dim ex As Exception = Nothing
		Try
			Console.WriteLine("Try")
		Catch ex
			Console.WriteLine("Catch")
		End Try
		Console.WriteLine(ex)
	End Sub

	Public Shared Sub TryCatchExistingLocalWhen()
		Dim ex As Exception = Nothing
		Try
			Console.WriteLine("Try")
		Catch ex When ex.Message IsNot Nothing
			Console.WriteLine("Catch")
		End Try
		Console.WriteLine(ex)
	End Sub

	Public Shared Sub TryCatchFinally()
		Try
			Console.WriteLine("Try")
		Catch ex As Exception
			Console.WriteLine(ex.Message)
		Finally
			Console.WriteLine("Finally")
		End Try
	End Sub

	Public Shared Sub TryCatchBareFinally()
		Try
			Console.WriteLine("Try")
		Catch
			Console.WriteLine("Catch")
		Finally
			Console.WriteLine("Finally")
		End Try
	End Sub

	Public Shared Sub TryCatchWhenFinally()
		Try
			Console.WriteLine("Try")
		Catch ex As Exception When Condition()
			Console.WriteLine(ex.Message)
		Finally
			Console.WriteLine("Finally")
		End Try
	End Sub

	Public Shared Sub TryMultipleCatch()
		Try
			Console.WriteLine("Try")
		Catch ex As FileNotFoundException
			Console.WriteLine(ex.FileName)
		Catch ex As IOException
			Console.WriteLine(ex.Message)
		Catch
			Console.WriteLine("Catch")
		End Try
	End Sub

	Public Shared Sub TryMultipleCatchFinally()
		Try
			Console.WriteLine("Try")
		Catch ex As FileNotFoundException
			Console.WriteLine(ex.FileName)
		Catch ex As IOException
			Console.WriteLine(ex.Message)
		Finally
			Console.WriteLine("Finally")
		End Try
	End Sub

	Public Shared Sub TryMultipleCatchWhen()
		Try
			Console.WriteLine("Try")
		Catch ex As IOException When Condition()
			Console.WriteLine(ex.Message)
		Catch When Condition()
			Console.WriteLine("Catch When")
		Catch ex As Exception
			Console.WriteLine(ex.Message)
		End Try
	End Sub

	Public Shared Sub EmptyTryFinally()
		Try
		Finally
			Console.WriteLine("Finally")
		End Try
	End Sub

	Public Shared Sub EmptyCatch()
		Try
			Console.WriteLine("Try")
		Catch
		End Try
	End Sub

	Public Shared Sub EmptyFinally()
		Try
			Console.WriteLine("Try")
		Catch ex As Exception
			Console.WriteLine(ex.Message)
		Finally
		End Try
	End Sub

	Public Shared Sub ExitTryInTry(b As Boolean)
		Try
			If b Then
				Exit Try
			End If
			Console.WriteLine("Try")
		Catch
			Console.WriteLine("Catch")
		End Try
		Console.WriteLine("End")
	End Sub

	Public Shared Sub ExitTryInCatch(b As Boolean)
		Try
			Console.WriteLine("Try")
		Catch
			If b Then
				Exit Try
			End If
			Console.WriteLine("Catch")
		End Try
		Console.WriteLine("End")
	End Sub

	Public Shared Sub ExitTryWithFinally(b As Boolean)
		Try
			If b Then
				Exit Try
			End If
			Console.WriteLine("Try")
		Finally
			Console.WriteLine("Finally")
		End Try
		Console.WriteLine("End")
	End Sub

	Public Shared Sub Rethrow()
		Try
			Console.WriteLine("Try")
		Catch ex As Exception
			Console.WriteLine(ex.Message)
			Throw
		End Try
	End Sub

	Public Shared Sub ThrowNew()
		Try
			Console.WriteLine("Try")
		Catch ex As Exception
			Throw New InvalidOperationException("Catch", ex)
		End Try
	End Sub

	Public Shared Function ReturnFromTry() As Integer
		Try
			Return 1
		Catch
			Return 2
		Finally
			Console.WriteLine("Finally")
		End Try
	End Function

	Public Shared Function ReturnAfterTry() As Integer
		Try
			Console.WriteLine("Try")
		Catch ex As Exception
			Console.WriteLine(ex.Message)
			Return -1
		End Try
		Return 0
	End Function

	Public Shared Sub NestedTryInTry()
		Try
			Try
				Console.WriteLine("Inner Try")
			Catch ex As IOException
				Console.WriteLine(ex.Message)
			End Try
		Catch ex As Exception
			Console.WriteLine(ex.Message)
		End Try
	End Sub

	Public Shared Sub NestedTryInCatch()
		Try
			Console.WriteLine("Try")
		Catch ex As Exception
			Try
				Console.WriteLine(ex.Message)
			Catch ex2 As Exception
				Console.WriteLine(ex2.Message)
			End Try
		End Try
	End Sub

	Public Shared Sub NestedTryInFinally()
		Try
			Console.WriteLine("Try")
		Finally
			Try
				Console.WriteLine("Inner Try")
			Catch
				Console.WriteLine("Inner Catch")
			End Try
		End Try
	End Sub

	Public Shared Sub TryInLoop()
		For i = 0 To 9
			Try
				If i = 5 Then
					Continue For
				End If
				If i = 8 Then
					Exit For
				End If
				Console.WriteLine(i)
			Catch
				Console.WriteLine("Catch")
			Finally
				Console.WriteLine("Finally")
			End Try
		Next
	End Sub
End Class

File: ICSharpCode.Decompiler.Tests/TestCases/VBPretty/VBTryCatchFinally.cs
using System;
using System.IO;

using Microsoft.VisualBasic.CompilerServices;

public class VBTryCatchFinally
{
	private static bool Condition()
	{
		return true;
	}

	public static void TryFinally()
	{
		try
		{
			Console.WriteLine("Try");
		}
		finally
		{
			Console.WriteLine("Finally");
		}
	}

	public static void TryCatchBare()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (Exception projectError)
		{
			ProjectData.SetProjectError(projectError);
			Console.WriteLine("Catch");
			ProjectData.ClearProjectError();
		}
	}

	public static void TryCatchVariable()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (Exception ex)
		{
			ProjectData.SetProjectError(ex);
			Exception ex2 = ex;
			Console.WriteLine(ex2.Message);
			ProjectData.ClearProjectError();
		}
	}

	public static void TryCatchSpecificType()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (IOException ex)
		{
			ProjectData.SetProjectError(ex);
			IOException ex2 = ex;
			Console.WriteLine(ex2.Message);
			ProjectData.ClearProjectError();
		}
	}

	public static void TryCatchUnusedVariable()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (InvalidOperationException ex)
		{
			ProjectData.SetProjectError(ex);
			InvalidOperationException ex2 = ex;
			Console.WriteLine("Catch");
			ProjectData.ClearProjectError();
		}
	}

	public static void TryCatchWhen()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (Exception projectError) when (((Func<bool>)delegate {
			// Could not convert BlockContainer to single expression
			ProjectData.SetProjectError(projectError);
			return Condition();
		}).Invoke())
		{
			Console.WriteLine("Catch");
			ProjectData.ClearProjectError();
		}
	}

	public static void TryCatchVariableWhen()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (Exception ex) when (((Func<bool>)delegate {
			// Could not convert BlockContainer to single expression
			ProjectData.SetProjectError(ex);
			return ex.Message != null;
		}).Invoke())
		{
			Console.WriteLine(ex.Message);
			ProjectData.ClearProjectError();
		}
	}

	public static void TryCatchSpecificTypeWhen()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (IOException ex) when (((Func<bool>)delegate {
			// Could not convert BlockContainer to single expression
			ProjectData.SetProjectError(ex);
			return Condition();
		}).Invoke())
		{
			Console.WriteLine(ex.Message);
			ProjectData.ClearProjectError();
		}
	}

	public static void TryCatchExistingLocal()
	{
		Exception value = null;
		try
		{
			Console.WriteLine("Try");
		}
		catch (Exception ex)
		{
			ProjectData.SetProjectError(ex);
			value = ex;
			Console.WriteLine("Catch");
			ProjectData.ClearProjectError();
		}
		Console.WriteLine(value);
	}

	public static void TryCatchExistingLocalWhen()
	{
		Exception value = null;
		try
		{
			Console.WriteLine("Try");
		}
		catch (Exception ex) when (((Func<bool>)delegate {
			// Could not convert BlockContainer to single expression
#if LEGACY_VBC
			value = ex;
			ProjectData.SetProjectError(ex);
#else
			ProjectData.SetProjectError(ex);
			value = ex;
#endif
			return ex.Message != null;
		}).Invoke())
		{
			Console.WriteLine("Catch");
			ProjectData.ClearProjectError();
		}
		Console.WriteLine(value);
	}

	public static void TryCatchFinally()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (Exception ex)
		{
			ProjectData.SetProjectError(ex);
			Exception ex2 = ex;
			Console.WriteLine(ex2.Message);
			ProjectData.ClearProjectError();
		}
		finally
		{
			Console.WriteLine("Finally");
		}
	}

	public static void TryCatchBareFinally()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (Exception projectError)
		{
			ProjectData.SetProjectError(projectError);
			Console.WriteLine("Catch");
			ProjectData.ClearProjectError();
		}
		finally
		{
			Console.WriteLine("Finally");
		}
	}

	public static void TryCatchWhenFinally()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (Exception ex) when (((Func<bool>)delegate {
			// Could not convert BlockContainer to single expression
			ProjectData.SetProjectError(ex);
			return Condition();
		}).Invoke())
		{
			Console.WriteLine(ex.Message);
			ProjectData.ClearProjectError();
		}
		finally
		{
			Console.WriteLine("Finally");
		}
	}

	public static void TryMultipleCatch()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (FileNotFoundException ex)
		{
			ProjectData.SetProjectError(ex);
			FileNotFoundException ex2 = ex;
			Console.WriteLine(ex2.FileName);
			ProjectData.ClearProjectError();
		}
		catch (IOException ex3)
		{
			ProjectData.SetProjectError(ex3);
			IOException ex4 = ex3;
			Console.WriteLine(ex4.Message);
			ProjectData.ClearProjectError();
		}
		catch (Exception projectError)
		{
			ProjectData.SetProjectError(projectError);
			Console.WriteLine("Catch");
			ProjectData.ClearProjectError();
		}
	}

	public static void TryMultipleCatchFinally()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (FileNotFoundException ex)
		{
			ProjectData.SetProjectError(ex);
			FileNotFoundException ex2 = ex;
			Console.WriteLine(ex2.FileName);
			ProjectData.ClearProjectError();
		}
		catch (IOException ex3)
		{
			ProjectData.SetProjectError(ex3);
			IOException ex4 = ex3;
			Console.WriteLine(ex4.Message);
			ProjectData.ClearProjectError();
		}
		finally
		{
			Console.WriteLine("Finally");
		}
	}

	public static void TryMultipleCatchWhen()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (IOException ex) when (((Func<bool>)delegate {
			// Could not convert BlockContainer to single expression
			ProjectData.SetProjectError(ex);
			return Condition();
		}).Invoke())
		{
			Console.WriteLine(ex.Message);
			ProjectData.ClearProjectError();
		}
		catch (Exception projectError) when (((Func<bool>)delegate {
			// Could not convert BlockContainer to single expression
			ProjectData.SetProjectError(projectError);
			return Condition();
		}).Invoke())
		{
			Console.WriteLine("Catch When");
			ProjectData.ClearProjectError();
		}
		catch (Exception ex2)
		{
			ProjectData.SetProjectError(ex2);
			Exception ex3 = ex2;
			Console.WriteLine(ex3.Message);
			ProjectData.ClearProjectError();
		}
	}

	public static void EmptyTryFinally()
	{
		try
		{
		}
		finally
		{
			Console.WriteLine("Finally");
		}
	}

	public static void EmptyCatch()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (Exception projectError)
		{
			ProjectData.SetProjectError(projectError);
			ProjectData.ClearProjectError();
		}
	}

	public static void EmptyFinally()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (Exception ex)
		{
			ProjectData.SetProjectError(ex);
			Exception ex2 = ex;
			Console.WriteLine(ex2.Message);
			ProjectData.ClearProjectError();
		}
#if !OPT || LEGACY_VBC
		finally
		{
		}
#endif
	}

	public static void ExitTryInTry(bool b)
	{
		try
		{
			if (!b)
			{
				Console.WriteLine("Try");
			}
		}
		catch (Exception projectError)
		{
			ProjectData.SetProjectError(projectError);
			Console.WriteLine("Catch");
			ProjectData.ClearProjectError();
		}
		Console.WriteLine("End");
	}

	public static void ExitTryInCatch(bool b)
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (Exception projectError)
		{
			ProjectData.SetProjectError(projectError);
			if (b)
			{
				ProjectData.ClearProjectError();
			}
			else
			{
				Console.WriteLine("Catch");
				ProjectData.ClearProjectError();
			}
		}
		Console.WriteLine("End");
	}

	public static void ExitTryWithFinally(bool b)
	{
		try
		{
			if (!b)
			{
				Console.WriteLine("Try");
			}
		}
		finally
		{
			Console.WriteLine("Finally");
		}
		Console.WriteLine("End");
	}

	public static void Rethrow()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (Exception ex)
		{
			ProjectData.SetProjectError(ex);
			Exception ex2 = ex;
			Console.WriteLine(ex2.Message);
			throw;
		}
	}

	public static void ThrowNew()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (Exception ex)
		{
			ProjectData.SetProjectError(ex);
			Exception innerException = ex;
			throw new InvalidOperationException("Catch", innerException);
		}
	}

	public static int ReturnFromTry()
	{
		int result;
		try
		{
			result = 1;
		}
		catch (Exception projectError)
		{
			ProjectData.SetProjectError(projectError);
			result = 2;
			ProjectData.ClearProjectError();
		}
		finally
		{
			Console.WriteLine("Finally");
		}
		return result;
	}

	public static int ReturnAfterTry()
	{
#if !(LEGACY_VBC && OPT)
		int result;
#endif
		try
		{
			Console.WriteLine("Try");
		}
		catch (Exception ex)
		{
			ProjectData.SetProjectError(ex);
			Exception ex2 = ex;
			Console.WriteLine(ex2.Message);
#if LEGACY_VBC && OPT
			int result = -1;
			ProjectData.ClearProjectError();
			return result;
		}
		return 0;
#else
			result = -1;
			ProjectData.ClearProjectError();
#if LEGACY_VBC
			goto IL_0032;
#elif OPT
			goto IL_0029;
#else
			goto IL_0031;
#endif
		}
		result = 0;
#if LEGACY_VBC
		goto IL_0032;
		IL_0032:
#elif OPT
		goto IL_0029;
		IL_0029:
#else
		goto IL_0031;
		IL_0031:
#endif
		return result;
#endif
	}

	public static void NestedTryInTry()
	{
		try
		{
			try
			{
				Console.WriteLine("Inner Try");
			}
			catch (IOException ex)
			{
				ProjectData.SetProjectError(ex);
				IOException ex2 = ex;
				Console.WriteLine(ex2.Message);
				ProjectData.ClearProjectError();
			}
		}
		catch (Exception ex3)
		{
			ProjectData.SetProjectError(ex3);
			Exception ex4 = ex3;
			Console.WriteLine(ex4.Message);
			ProjectData.ClearProjectError();
		}
	}

	public static void NestedTryInCatch()
	{
		try
		{
			Console.WriteLine("Try");
		}
		catch (Exception ex)
		{
			ProjectData.SetProjectError(ex);
			Exception ex2 = ex;
			try
			{
				Console.WriteLine(ex2.Message);
			}
			catch (Exception ex3)
			{
				ProjectData.SetProjectError(ex3);
				Exception ex4 = ex3;
				Console.WriteLine(ex4.Message);
				ProjectData.ClearProjectError();
			}
			ProjectData.ClearProjectError();
		}
	}

	public static void NestedTryInFinally()
	{
		try
		{
			Console.WriteLine("Try");
		}
		finally
		{
			try
			{
				Console.WriteLine("Inner Try");
			}
			catch (Exception projectError)
			{
				ProjectData.SetProjectError(projectError);
				Console.WriteLine("Inner Catch");
				ProjectData.ClearProjectError();
			}
		}
	}

	public static void TryInLoop()
	{
		int num = 0;
		do
		{
			try
			{
				switch (num)
				{
					case 5:
						break;
					case 8:
						return;
					default:
						Console.WriteLine(num);
						break;
				}
			}
			catch (Exception projectError)
			{
				ProjectData.SetProjectError(projectError);
				Console.WriteLine("Catch");
				ProjectData.ClearProjectError();
			}
			finally
			{
				Console.WriteLine("Finally");
			}
			num = checked(num + 1);
		} while (num <= 9);
	}
}
